File: Assets/Plugins/Android/src/com/playphone/multinet/unity/MNVItemsProviderUnity.java
package com.playphone.multinet.unity;

import com.playphone.multinet.*;
import com.unity3d.player.UnityPlayer;
import com.playphone.multinet.providers.*;

public class MNVItemsProviderUnity {
    public static void shutdown() {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.getVItemsProvider().shutdown();
            }
        });
    }

    public static String getGameVItemsList() {
        MNUnity.MARK();

        return MNUnity.serializer.serialize(MNDirect.getVItemsProvider().getGameVItemsList());
    }

    public static String findGameVItemById(final int vItemId) {
        MNUnity.MARK();

        return MNUnity.serializer.serialize(MNDirect.getVItemsProvider().findGameVItemById(vItemId));
    }

    public static boolean isGameVItemsListNeedUpdate() {
        MNUnity.MARK();

        return MNDirect.getVItemsProvider().isGameVItemsListNeedUpdate();
    }

    public static void doGameVItemsListUpdate() {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.getVItemsProvider().doGameVItemsListUpdate();
            }
        });
    }

    public static void reqAddPlayerVItem(final int vItemId, final long count, final long clientTransactionId) {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.getVItemsProvider().reqAddPlayerVItem(vItemId, count, clientTransactionId);
            }
        });
    }

    public static void reqAddPlayerVItemTransaction(final String transactionVItems, final long clientTransactionId) {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.getVItemsProvider().reqAddPlayerVItemTransaction(MNUnity.serializer.deserialize(transactionVItems,MNVItemsProvider.TransactionVItemInfo[].class), clientTransactionId);
            }
        });
    }

    public static void reqTransferPlayerVItem(final int vItemId, final long count, final long toPlayerId, final long clientTransactionId) {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.getVItemsProvider().reqTransferPlayerVItem(vItemId, count, toPlayerId, clientTransactionId);
            }
        });
    }

    public static void reqTransferPlayerVItemTransaction(final String transactionVItems, final long toPlayerId, final long clientTransactionId) {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.getVItemsProvider().reqTransferPlayerVItemTransaction(MNUnity.serializer.deserialize(transactionVItems,MNVItemsProvider.TransactionVItemInfo[].class), toPlayerId, clientTransactionId);
            }
        });
    }

    public static String getPlayerVItemList() {
        MNUnity.MARK();

        return MNUnity.serializer.serialize(MNDirect.getVItemsProvider().getPlayerVItemList());
    }

    public static long getPlayerVItemCountById(final int vItemId) {
        MNUnity.MARK();

        return MNDirect.getVItemsProvider().getPlayerVItemCountById(vItemId);
    }

    public static String getVItemImageURL(final int vItemId) {
        MNUnity.MARK();

        return MNDirect.getVItemsProvider().getVItemImageURL(vItemId);
    }

    public static long getNewClientTransactionId() {
        MNUnity.MARK();

        return MNDirect.getVItemsProvider().getNewClientTransactionId();
    }

    protected static class MNVItemsProviderEventHandler implements com.playphone.multinet.providers.MNVItemsProvider.IEventHandler {

        @Override
        public void onVItemsListUpdated() {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_onVItemsListUpdated");
                }
            });
        }

        @Override
        public void onVItemsTransactionCompleted(final MNVItemsProvider.TransactionInfo transaction) {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_onVItemsTransactionCompleted",transaction);
                }
            });
        }

        @Override
        public void onVItemsTransactionFailed(final MNVItemsProvider.TransactionError error) {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_onVItemsTransactionFailed",error);
                }
            });
        }
    }

    private static MNVItemsProviderEventHandler registeredEventHandler = null;
    public static synchronized boolean registerEventHandler() {
        MNUnity.MARK();

        if (MNDirect.getVItemsProvider() == null) {
            MNUnity.ELog("MNVItemsProvider is not ready. Use MNDirect\'s sessionReady event for adding your delegates.");
            return false;
        }

        if (registeredEventHandler == null) {
            registeredEventHandler = new MNVItemsProviderUnity.MNVItemsProviderEventHandler();
            MNDirect.getVItemsProvider().addEventHandler(registeredEventHandler);
        }

        return true;
    }
    
    public static synchronized boolean unregisterEventHandler() {
        MNUnity.MARK();

        if (MNDirect.getVItemsProvider() == null) {
            return false;
        }

        if (registeredEventHandler != null) {
            MNDirect.getVItemsProvider().removeEventHandler(registeredEventHandler);
            registeredEventHandler = null;
        }

        return true;
    }
}

File: Assets/Plugins/iOS/PlayPhoneMultiNet/unity/providers/MNVItemsProviderUnity.m
//
//  MNVItemsProviderUnity.m
//  PlayPhone SDK for Unity
//
//  Copyright (c) 2012 PlayPhone Inc. All rights reserved.
//

#import "MNDirect.h"
#import "MNVItemsProvider.h"

#import "MNVItemsProviderUnity.h"

@interface MNVItemsProviderUnity : NSObject <MNVItemsProviderDelegate>
+ (id)shared;
- (void)onVItemsListUpdated;
- (void)onVItemsTransactionCompleted:(MNVItemsTransactionInfo*)transaction;
- (void)onVItemsTransactionFailed:(MNVItemsTransactionError*)error;
@end

static MNVItemsProviderUnity *MNVItemsProviderUnityInstance = nil;

@implementation MNVItemsProviderUnity

+(id)shared {
    if (MNVItemsProviderUnityInstance == nil) {
        MNVItemsProviderUnityInstance = [[MNVItemsProviderUnity alloc]init];
    }
    
    return MNVItemsProviderUnityInstance;
}

- (void)onVItemsListUpdated {
    [MNUnity callUnityFunction:@"MNUM_onVItemsListUpdated" withParams:nil];
}
- (void)onVItemsTransactionCompleted:(MNVItemsTransactionInfo*)transaction {
    [MNUnity callUnityFunction:@"MNUM_onVItemsTransactionCompleted" withParams:(transaction == nil ? [NSNull null] : transaction), nil];
}
- (void)onVItemsTransactionFailed:(MNVItemsTransactionError*)error {
    [MNUnity callUnityFunction:@"MNUM_onVItemsTransactionFailed" withParams:(error == nil ? [NSNull null] : error), nil];
}
@end

EXTERN_C
const char* _MNVItemsProvider_GetGameVItemsList() {
    MARK;

    return MakeStringCopy([[[MNUnity serializer]serialize:[[MNDirect vItemsProvider] getGameVItemsList]] UTF8String]);
}

EXTERN_C
const char* _MNVItemsProvider_FindGameVItemById(int vItemId) {
    MARK;

    return MakeStringCopy([[[MNUnity serializer]serialize:[[MNDirect vItemsProvider] findGameVItemById:vItemId]] UTF8String]);
}

EXTERN_C
bool _MNVItemsProvider_IsGameVItemsListNeedUpdate() {
    MARK;

    return [[MNDirect vItemsProvider] isGameVItemsListNeedUpdate];
}

EXTERN_C
void _MNVItemsProvider_DoGameVItemsListUpdate() {
    MARK;

    [[MNDirect vItemsProvider] doGameVItemsListUpdate];
}

EXTERN_C
void _MNVItemsProvider_ReqAddPlayerVItem(int vItemId, long long count, long long clientTransactionId) {
    MARK;

    [[MNDirect vItemsProvider] reqAddPlayerVItem:vItemId count:count andClientTransactionId:clientTransactionId];
}

EXTERN_C
void _MNVItemsProvider_ReqAddPlayerVItemTransaction(const char* transactionVItems, long long clientTransactionId) {
    MARK;

    [[MNDirect vItemsProvider] reqAddPlayerVItemTransaction:[[MNUnity serializer] deserializeArray:[MNTransactionVItemInfo class] fromJson:[NSString stringWithUTF8String:transactionVItems]] andClientTransactionId:clientTransactionId];
}

EXTERN_C
void _MNVItemsProvider_ReqTransferPlayerVItem(int vItemId, long long count, long long toPlayerId, long long clientTransactionId) {
    MARK;

    [[MNDirect vItemsProvider] reqTransferPlayerVItem:vItemId count:count toPlayer:toPlayerId andClientTransactionId:clientTransactionId];
}

EXTERN_C
void _MNVItemsProvider_ReqTransferPlayerVItemTransaction(const char* transactionVItems, long long toPlayerId, long long clientTransactionId) {
    MARK;

    [[MNDirect vItemsProvider] reqTransferPlayerVItemTransaction:[[MNUnity serializer] deserializeArray:[MNTransactionVItemInfo class] fromJson:[NSString stringWithUTF8String:transactionVItems]] toPlayer:toPlayerId andClientTransactionId:clientTransactionId];
}

EXTERN_C
const char* _MNVItemsProvider_GetPlayerVItemList() {
    MARK;

    return MakeStringCopy([[[MNUnity serializer]serialize:[[MNDirect vItemsProvider] getPlayerVItemList]] UTF8String]);
}

EXTERN_C
long long _MNVItemsProvider_GetPlayerVItemCountById(int vItemId) {
    MARK;

    return [[MNDirect vItemsProvider] getPlayerVItemCountById:vItemId];
}

EXTERN_C
const char* _MNVItemsProvider_GetVItemImageURL(int vItemId) {
    MARK;

    NSURL *resultUrl = [[MNDirect vItemsProvider] getVItemImageURL:vItemId];

    if (resultUrl == nil) {
        return NULL;
    }

    return MakeStringCopy([[resultUrl absoluteString] UTF8String]);
}

EXTERN_C
long long _MNVItemsProvider_GetNewClientTransactionId() {
    MARK;

    return [[MNDirect vItemsProvider] getNewClientTransactionId];
}

EXTERN_C
bool _MNVItemsProvider_RegisterEventHandler(){
    MARK;

    if ([MNDirect vItemsProvider] == nil) {
        return false;
    }

    [[MNDirect vItemsProvider] addDelegate:[MNVItemsProviderUnity shared]];

    return true;
}

EXTERN_C
bool _MNVItemsProvider_UnregisterEventHandler(){
    MARK;

    if ([MNDirect vItemsProvider] == nil) {
        return false;
    }

    [[MNDirect vItemsProvider] removeDelegate:[MNVItemsProviderUnity shared]];

    return true;
}


